server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+        TStorageGroupSchema storageGroupSchema = new TStorageGroupSchema();
+        storageGroupSchema.setName(storageGroupName);
+        TSetStorageGroupReq req = new TSetStorageGroupReq(storageGroupSchema);
+        TSStatus tsStatus = client.setStorageGroup(req);
+        if (TSStatusCode.SUCCESS_STATUS.getStatusCode() == tsStatus.getCode()) {
+          successFullyCreatedStorageGroup.add(storageGroupName);
+        } else {
+          updateStorageCache(successFullyCreatedStorageGroup);
+          throw new RuntimeException(new IoTDBException(tsStatus.message, tsStatus.code));
+        }
+      }
+      updateStorageCache(storageGroupNamesNeedCreated);
+    }
+  }
+
+  /** update storage group cache */
+  public void updateStorageCache(Set<String> storageGroupNames) {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      storageGroupCache.addAll(storageGroupNames);
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  /** invalid storage group cache after delete */
+  public void invalidStorageGroupCache(List<String> storageGroupNames) {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      for (String storageGroupName : storageGroupNames) {
+        storageGroupCache.remove(storageGroupName);
+      }
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  /** invalid all storage group */
+  public void invalidAllStorageGroupCache() {
+    storageGroupCacheLock.writeLock().lock();
+    try {
+      storageGroupCache.clear();
+    } finally {
+      storageGroupCacheLock.writeLock().unlock();
+    }
+  }
+
+  // endregion
+
+  // region replicaSet cache
+  /** get regionReplicaSet from consensusGroupId */
+  public TRegionReplicaSet getRegionReplicaSet(TConsensusGroupId consensusGroupId) {
+    TRegionReplicaSet regionReplicaSet;
+    if (!groupIdToReplicaSetMap.containsKey(consensusGroupId)) {
+      try (ConfigNodeClient client =
+          configNodeClientManager.borrowClient(ConfigNodeInfo.partitionRegionId)) {
+        TRegionRouteMapResp resp = client.getLatestRegionRouteMap();
+        if (TSStatusCode.SUCCESS_STATUS.getStatusCode() == resp.getStatus().getCode()) {
+          updateGroupIdToReplicaSetMap(resp.getTimestamp(), resp.getRegionRouteMap());
+        }
+        if (!groupIdToReplicaSetMap.containsKey(consensusGroupId)) {
+          throw new RuntimeException(
+              "Failed to get replicaSet of consensus group[id= " + consensusGroupId + "]");
+        }
+      } catch (IOException | TException e) {
+        throw new StatementAnalyzeException(
+            "An error occurred when executing getRegionReplicaSet():" + e.getMessage());
+      }
+    }
+    regionReplicaSet = groupIdToReplicaSetMap.get(consensusGroupId);
+    if (null == regionReplicaSet) {
+      throw new RuntimeException(
+          "ReplicaSet of consensus group[id= " + consensusGroupId + "] is null.");
+    }
+    return regionReplicaSet;
```

Comment:
The RuntimeException might cause a NPE in upper layer. We should notify this in the function annotation in order to reduce the communication costs that might be inccured later.

Reply:
Get it ~